ROW Inspector Prioritization ETL

Charlie-Henry opened the pull request:
Closes https://github.com/cityofaustin/atd-data-tech/issues/19680

This PR adds a new `metric` script which generates a priority score for ROW permits based on several performance measures . This was originally piloted by @johnclary 5 years ago over in [atd-amanda](https://github.com/cityofaustin/atd-amanda/tree/master/inspector-dash). This (I believe) faithfully copies every performance measure from that original work and includes the rest of the ETL to retrieve data from AMANDA (the original work was done "manually" with CSV extracts from AMANDA).

johnclary commented on README.md:
```diff
@@ -21,37 +21,44 @@ AMANDA is the backend system that underlies the [Austin Build + Connect](https:/
+
+`inspector_prioritization.py` "scores" permits based on several metrics to rank permits based on a prioritization for ROW inspectors.
+
+`python metrics/inspector_prioritization.py`
```
it would be good to document that this script requires that the S3-loading scripts have been run beforehand to ensure it's using fresh data

johnclary commented on README.md:
```diff
@@ -21,37 +21,44 @@ AMANDA is the backend system that underlies the [Austin Build + Connect](https:/
 Then, provide the environment variables described in env_template to the docker image:
 
-`$ docker run -it --env-file env_file dts-right-of-way-reporting:production /bin/bash` 
+`docker run -it --env-file env_file dts-right-of-way-reporting:production /bin/bash` 
```
this command is missing the `atddocker` repo name. so it should be: 

```
docker run -it --env-file env_file atddocker/dts-right-of-way-reporting:production /bin/bash
```

johnclary replied:
ooh and it would be nice for us local devvies to have the volume mountadded in 😇 

```
docker run -it --env-file env_file -v "$(pwd):/app" atddocker/dts-right-of-way-reporting:production /bin/bash
```

johnclary commented on metrics/inspector_prioritization.py:
```diff
@@ -0,0 +1,274 @@
+import boto3
+import pandas as pd
+import numpy as np
+from arcgis.gis import GIS
```
edit: i was not able to get this running :/

ignore this 👇 

i had to install `arcgis` in order to get this running—looks like it's missing from the requirements.

annoyingly, i could only install it after pulling the amd64 version of the image. so this:
```shell
docker run -it --env-file env_file --platform linux/amd64 -v "$(pwd):/app" atddocker/dts-right-of-way-reporting:production /bin/bash
```

then this

```shell
pip install arcgis
```

Charlie-Henry replied:
@johnclary I was able to find a patch for the dockerfile to get it to build for me

johnclary commented on metrics/inspector_prioritization.py:
```diff
@@ -0,0 +1,274 @@
+    permits["count_segments"] = permits["count_segments"].fillna(0)
+
+    # 10 points for permits with more than 1 segment, 5 points otherwise:
+    permits["count_segment_scoring"] = np.where(permits["count_segments"] > 1, 10, 5)
```
i don't know if this is really practical to implement, but it could be nice to collect all these magic numbers used for scoring into named vars at the top of the script, or a dict. that would make life easier the next time we need to tweak the scoring parameters.

johnclary commented on metrics/inspector_prioritization.py:
```diff
@@ -0,0 +1,274 @@
+        yield data[i : i + batch_size]
+
+
+def retrieve_road_segment_data(segments, gis):
```
would be nice to have a little more documentation here. looks like you're grabbing the geometry of every street segment that has a permit so that you can assign it ROW inspector zones.

johnclary commented on metrics/inspector_prioritization.py:
```diff
@@ -0,0 +1,274 @@
+    if response.features:
+        # It is possible a segment could intersect multiple zones, but we only take the first one.
+        return response.features[0].attributes["ROW_INSPECTOR_ZONE_ID"]
+    return None
```
this all seems to be working fine! if this needed to run at a high frequency, or if the AGOL API were giving us trouble, it would be cool to write an ETL that makes a big ole lookup table of all the segments by inspector zone and dapcz intersection. this ETL could pull from the lookup table, since it probably would not need to be updated very often.

Charlie-Henry replied:
@johnclary see [here](https://datahub.austintexas.gov/Transportation-and-Mobility/Street-Centerlines-with-ROW-Inspector-Zones/nt54-hvix/about_data)!

johnclary reviewed (commented):
Code looks good! I haven't been able to test this because I can't get the python env working on my Mac—with pip or docker.

TracyLinder reviewed (commented):
I'm by no means qualified to review a PR so I don't have anything of import to say. It's cool to see this side of the dev world though and y'all should invite me more often. ☺️

frankhereford commented on Dockerfile:
```diff
@@ -9,5 +9,6 @@ COPY . /app
 RUN chmod -R 755 /app/*
 
 # # Proceed to install the requirements...do
```
Certainly nothing but housekeeping, but you have double commented this line out.

frankhereford commented on Dockerfile:
```diff
@@ -9,5 +9,6 @@ COPY . /app
 
 # # Proceed to install the requirements...do
+RUN apt-get install libkrb5-dev -y
 RUN apt-get --allow-releaseinfo-change update
```
I would reverse the order of these lines. The update is refreshing the system's cache of the upstream package repositories, and the install relies on that information being up to date. 

Additionally, there is an update on line 3, and it'd be cool to consolidate the Dockerfile to have one less line, yielding one less layer in the image's filesystem.

frankhereford commented on README.md:
```diff
@@ -21,37 +21,46 @@ AMANDA is the backend system that underlies the [Austin Build + Connect](https:/
 This repo can be used with a docker container. You can either build it yourself with:
 
-`$ docker build . -t dts-right-of-way-reporting:production`
+`docker build . -t dts-right-of-way-reporting:production`
```
When people are building it for themselves, I think it would be cool to suggest they not use the `:production` tag. Maybe `:testing` or `:local`? The idea is to make it a little harder to conflate something built in a dev environment as the production image and accidentally pushing it up to dockerhub.

You don't have the `atddocker` prefix on here, which is helpful though, so this isn't the end of the world -- really more of a way to suggest safety-first habits to folks.

frankhereford commented on requirements.txt:
Since there's a modification to the requirements here -- can we bring the rest up to date? All but two have had a minor version release and none have had a major version release, so I'm hopeful that's not too much. Here's a proposed revision, current at the time of writing. The commented out versions are the latest and greatest on pypi, and they work, with one exception (`pandas`).

```
pandas==2.1.* # 2.2.3, but arcgis demands < 2.2 for pandas
boto3==1.35.* # 1.35.9
sodapy==2.2.* # 2.2.0
smartsheet-python-sdk==3.0.*
pytz==2024.2 # 2024.2
oracledb==2.5.* # 2.5.1
arcgis==2.4.* # 2.4.0
```

frankhereford replied:
@Charlie-Henry 🤦‍♂️ Sorry - I committed onto your branch. This was a file I was building up to include in a comment, and I just whale-failed on it and pushed my draft up into your feature branch. I'm about to undo this in a second commit. Sorry about that!

https://github.com/cityofaustin/dts-right-of-way-reporting/pull/4/commits/09ae3cfdd1a4900f381949415a908fd14f878c62

frankhereford commented on Dockerfile:
```diff
@@ -9,5 +9,6 @@ COPY . /app
 RUN chmod -R 755 /app/*
```
I'm unsure why we're setting everything in this broad file permission policy here. That's read, write and execute for everything in the folder, even things which are non-executable like the requirements file or the README, and then read and write for others in the group and then all others in the system at all. Given that it's a docker container, there is no real need to be concerned with anything except the root user, who could read, write and execute anything anyway. 

I think this line may be vestigial from some Dockerfile way back when, and it's unclear what purpose it serves here to me. Can we strike it?

Charlie-Henry replied:
I struck this, thanks Frank!

frankhereford reviewed (changes requested):
Thank you for this Charlie; there's a lot in here, to say the least. 

This built and ran without any issue from me. To other reviewers, if you want an environment file, LMK and I'll `wormhole` you one. It takes a second to build one to test this.

```
root@acd557b492f0:/app# python ./amanda/amanda_to_s3.py --query row_inspector_permit_list
2025-01-07 23:01:40,720 INFO: Executing query: row_inspector_permit_list
2025-01-07 23:02:41,580 INFO: Uploading 2479 rows to S3

root@acd557b492f0:/app# python amanda/amanda_to_s3.py --query row_inspector_segment_list
2025-01-07 23:03:01,334 INFO: Executing query: row_inspector_segment_list
2025-01-07 23:03:08,706 INFO: Uploading 4897 rows to S3

root@acd557b492f0:/app# python metrics/inspector_prioritization.py
/usr/local/lib/python3.10/site-packages/dask/dataframe/__init__.py:49: FutureWarning: 
Dask dataframe query planning is disabled because dask-expr is not installed.

You can install it with `pip install dask[dataframe]` or `conda install dask`.
This will raise in a future version.

  warnings.warn(msg, FutureWarning)
2025-01-07 23:03:30,635 INFO: 2479 Permits retrieved from S3
2025-01-07 23:03:30,739 INFO: 4897 Segments retrieved from S3
2025-01-07 23:03:33,798 INFO: Retrieving road segment data from AGOL...
/app/metrics/inspector_prioritization.py:123: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  primary_segments.drop_duplicates(subset=["SEGMENT_ID"], inplace=True)
2025-01-07 23:06:40,100 INFO: Scoring permits based on road segments data
2025-01-07 23:06:40,182 INFO: Replacing data in Socrata dataset: datahub.austintexas.gov/d/cavh-ec9p
2025-01-07 23:06:41,691 INFO: {'Errors': 0, 'Rows Created': 2479, 'Rows Updated': 0, 'Rows Deleted': 0}
root@acd557b492f0:/app# 
```

There are some warnings in the output that are going to turn into exceptions at some point that I want to make sure you're aware of.

---

This is shippable, but I do want to ask for a couple of changes -- but the world sure will keep spinning if we have to roll these off into issues. 

The things I'm most interested in are concerning the `Dockerfile` and the `requirements.txt`.  

In the `Dockerfile`, there are some out-of-order `apt` commands and a `chmod` that I think should get pitched. I'm curious if you agree. 

In the `requirements.txt`, since we're pinning the libraries, can we avoid as much tech debt as possible by pinning the most recent versions? I have a proposed revision in a comment below that is what I tested with, successfully. 

Other than that -- the real meat and potatoes of this is great - thanks again!

Charlie-Henry commented:
I've added a new script `roadway_segment_tagging.py` that populates a socrata [dataset](https://datahub.austintexas.gov/Transportation-and-Mobility/Street-Centerlines-with-ROW-Inspector-Zones/nt54-hvix/about_data) with the street segment inspector zones and DAPCZ segments. Running it should do nothing as it searches for segments with updates in the last 30 days, and there are none. If you want to test you can change line 89 to 180 days to see it working.

Otherwise, I don't see any ✅ 's from y'all so I hope I addressed most of your comments @frankhereford @johnclary

johnclary commented on docs/row_inspector_scoring.png:
this is such a nice visual 🙏

johnclary commented on metrics/roadway_segment_tagging.py:
very cool! would you mind adding some docs about this in the README? 

it might be good to link to the relevant Airflow DAG(s) for an example of what order all these scripts need to run in? or just write that out in the readme.

johnclary reviewed (approved):
It took 25 minutes to build the image locally, but it built! Thank you Charlie and Frank for sorting that tout.

The updated python scripts look great. I ran `inspector_prioitization` without issue and it ran very quickly. Nice!!!!!

chiaberry reviewed (approved):
I was able to run the scripts, the results of my third one was:

```
2025-01-22 16:21:57,561 INFO: Replacing data in Socrata dataset: datahub.austintexas.gov/d/cavh-ec9p
2025-01-22 16:21:59,127 INFO: {'Errors': 0, 'Rows Created': 2566, 'Rows Updated': 0, 'Rows Deleted': 0}
```

which is maybe 100 rows less than before i ran the script?

Charlie-Henry commented:
> I was able to run the scripts, the results of my third one was:
> 
> ```
> 2025-01-22 16:21:57,561 INFO: Replacing data in Socrata dataset: datahub.austintexas.gov/d/cavh-ec9p
> 2025-01-22 16:21:59,127 INFO: {'Errors': 0, 'Rows Created': 2566, 'Rows Updated': 0, 'Rows Deleted': 0}
> ```
> 
> which is maybe 100 rows less than before i ran the script?

Thanks Chia! I think this can be expected to go up or down as the number of active permits does fluctuate.
